File: PS-DP/계단 오르기.py
import sys
input = sys.stdin.readline
n = int(input())
arr = []
dp = []
for i in range(n):
    arr.append(int(input()))
    if i == 0:
        dp.append(arr[0])
    elif i == 1:
        dp.append(arr[0]+arr[1])
    elif i == 2:
        dp.append(max(arr[0] + arr[2], arr[1] + arr[2]))
    elif i == 3:
        dp.append(max(dp[1] + arr[3], arr[0] + arr[2] + arr[3]))
    else:
        dp.append(max(dp[i-2] + arr[i], arr[i] + arr[i-1] + dp[i-3]))
print(dp[n-1])

      

File: PS-DP/계단 오르기.cpp
#include<iostream>
#include<vector>
#include<queue>
using namespace std;

int n;
int arr[301] = {};
int dp[300][2];
int main() {
	cin >> n;
	for (int i = 1; i <= n; i++) {
		cin >> arr[i];
	}
	dp[0][1] = dp[0][2] = 0;
	dp[1][1] = dp[1][2] = arr[1];
	for (int i = 2; i <= n; i++) {
		dp[i][1] = dp[i - 1][2] + arr[i];
		dp[i][2] = max(dp[i - 2][1], dp[i - 2][2]) + arr[i];
	}
	cout << max(dp[n][1], dp[n][2]);
	return 0;
}